ocr/readingorder.py

```diff
@@ -1,212 +1,214 @@
+            bottom_of_first_line=line.rect.y1,
+            top_of_last_line=self.top_of_last_line
+        )
+
+    def is_interrupted_by(self, rect: pymupdf.Rect) -> bool:
+        y_middle = (rect.y0 + rect.y1) / 2
+        return rect.intersects(self.rect) and self.bottom_of_first_line < y_middle < self.top_of_last_line
+
+    def can_be_extended_by(self, geometry: ReadingOrderGeometry) -> bool:
+        column_width = self.rect.width
+        min_x = self.rect.x0 - 10 - 0.1 * column_width
+        max_x = self.rect.x1 + 10 + 0.1 * column_width
+        return (
+            geometry.y_middle > self.top_of_last_line and  # below
+            geometry.rect.y0 - self.rect.y1 < self.rect.height and  # not too far below
+            geometry.rect.x0 > min_x and
+            geometry.rect.x1 < max_x and
+            x_overlap(self.rect, geometry.rect) > 0.8 * geometry.rect.width
         )
 
+    def is_accurately_extended_by(self, geometry: ReadingOrderGeometry) -> bool:
+        return self.can_be_extended_by(geometry) and x_overlap(self.rect, geometry.rect) > 0.8 * self.rect.width and (
+            self.rect.y1 < geometry.rect.y1  #strictly below
+        )
 
-def overlaps(line, line2) -> bool:
-    vertical_margin = 15
-    ref_rect = pymupdf.Rect(line.rect.x0, line.rect.y0 - vertical_margin, line.rect.x1, line.rect.y1 + vertical_margin)
-    return ref_rect.intersects(line2.rect)
-
-
-def adjacent_lines(lines: list[TextLine]) -> list[set[int]]:
-    result = [set() for _ in lines]
-    for index, line in enumerate(lines):
-        for index2, line2 in enumerate(lines):
-            if index2 > index:
-                if overlaps(line, line2):
-                    result[index].add(index2)
-                    result[index2].add(index)
-    return result
-
-
-def apply_transitive_closure(data: list[set[int]]) -> bool:
-    found_new_relation = False
-    for index, adjacent_indices in enumerate(data):
-        new_adjacent_indices = set()
-        for adjacent_index in adjacent_indices:
-            new_adjacent_indices.update(
-                new_index
-                for new_index in data[adjacent_index]
-                if new_index not in data[index]
+    @classmethod
+    def current_column(
+            cls,
+            current_line: TextLineReadingOrder,
+            preceding_lines: list[TextLineReadingOrder],
+            all_lines: set[TextLineReadingOrder]
+    ) -> "ReadingOrderColumn":
+        other_lines = all_lines.copy()
+        other_lines.remove(current_line)
+        column = ReadingOrderColumn(
+            rect=current_line.geometry.rect,
+            bottom_of_first_line=current_line.geometry.rect.y1,
+            top_of_last_line=current_line.geometry.rect.y0
+        )
+        accurate_extension_count = sum(
+            1 for line in other_lines if column.is_accurately_extended_by(line.geometry)
+        )
+        for line in preceding_lines[::-1]:
```

Comment:
```suggestion
        for line in reversed(preceding_lines):
```
do we need to make sure that preceding lines are sorted by y0? or is that unnecessary?

Reply:
No, we actually don't want to sort by y0, we want to reversely follow the reading order as it was already established up to this point. I'll add a comment to clarify.

Reply:
Ah, got it! That makes sense, I didn't realize when writing the comment that preceding_lines always comes from current_block[:-1] in sort_lines, so the order is already consistent with reading order. Thanks for the clarification, I think its already clear that we follow the preceding lines reversely!